### PR TITLE
new tests to verify streaming capabilities of itkResampleImageFilter

### PR DESCRIPTION
**Pull request by romangrothausmann**
PR only for the tests that are expected to succeed with #82 (and to fail currently) in order to check their outcome without the changes to itkResampleImageFilter #82  (which currently appear to effect other tests like itkMultiResolutionPyramidImageFilterWithResampleFilterTest as well)
According to https://github.com/InsightSoftwareConsortium/ITK/pull/84#issuecomment-450907156, history squashed into single ENH-commit rebase on to master.

creation history:
1. plain copy of itkResampleImageTest.cxx and its cmake
2. employ itkResampleImageTest2s, still the same as itkResampleImageTest2
3. write MHAs (which support streaming) instead of PNGs
4. demand streaming and employ where possible
5. verify streaming for linear case
6. itkResampleImageTest7.cxx for comparing output of stream driven imaging (SDI) and without SDI
7. mark resample as modified to enforce re-execution of filter chain

### CODE REVIEW

**Comment by romangrothausmann** [comment]
As some changes were done on the tests to conform to the ITK5 standards, I guess `itkResampleImageTest2s.cxx` and `itkResampleImageTest7.cxx` should be re-done to incorporate the changes introduced for ITK5, now that the PR is based on master instead of v4.13.1?

File: Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    std::cout << "Usage: " << argv[0]
+              << " noParams" << std::endl;
+    return EXIT_FAILURE;
+    }
+
+  const unsigned int NDimensions = 2;
+
+  typedef unsigned char                          PixelType;
+  typedef itk::Image<PixelType, NDimensions>     ImageType;
+  typedef ImageType::IndexType         ImageIndexType;
```
Comment (dzenanz) [inline Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx]: The alignment is off for this block of declarations

**Comment by romangrothausmann** [comment]
Re-did the commit based on other the other test-files used now take from https://github.com/InsightSoftwareConsortium/ITK/tree/b2884129bf6db53cb102137d81e7176e893c5e5f
The corresponding branch that is not squashed is: https://github.com/romangrothausmann/ITK/commits/test_RSI-SDI_master_incr

File: Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx (hunk 1 above)
Comment (romangrothausmann) [reply Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx]: Should be resolved with new commit

**Comment by romangrothausmann** [comment]
New push contains changes as suggested by @jhlegarreta in https://github.com/InsightSoftwareConsortium/ITK/pull/84#pullrequestreview-188754779 together with a fixed typo.

**Comment by romangrothausmann** [comment]
It is not clear to me why the circleci job failed. If I'm not mistaken, both new tests passed according to the logs.

**Comment by romangrothausmann** [comment]
> It is not clear to me why the circleci job failed. If I'm not mistaken, both new tests passed according to the logs.

For clarification: the 2 new tests should fail for current master, therefore I added the expectation to `WILL_FAIL` https://github.com/romangrothausmann/ITK/commit/6b59a58c6ca8e0576a1e48d6dad00a22f093b4c3 (with #82 the tests should succeed so with #82 the expectation should be changed then).

However, the tests seem to succeed currently, so they are not yet doing what they are meant to do. What could be missing? @blowekamp could you take a look?

**Comment by thewtex** [comment]
Thanks for pushing this forward @romangrothausmann .  I will take a look :eyes:

File: Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  numStreamDiv= 8; // split into numStream pieces for streaming.
+  resample->Modified(); // enforce re-execution even though nothing of filter changed
+  streamer->SetNumberOfStreamDivisions(numStreamDiv);
+  TRY_EXPECT_NO_EXCEPTION( streamer->Update() );
+
+  if (!monitor->VerifyAllInputCanStream(numStreamDiv))
+    {
+    std::cout << "Streaming failed to execute as expected!" << std::endl;
+    std::cout << monitor;
+    return EXIT_FAILURE;
```
Comment (jhlegarreta) [inline Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx]: I'd suggest adding a
```
std::cerr << "Test failed!" << std::endl;
```
message and redirecting the output to `std::cerr` for the sake of consistency.

Comment (thewtex) [reply Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx]: Done.

**Comment by thewtex** [comment]
Taking a look, the `PipelineMonitorImageFilter` needs to be in a different position in the pipeline, i.e.,

reader -> monitor -> resample -> writerOrStreamingFilter

This way, if `resample` is streaming, it will only request smaller regions from `monitor`, and `monitor` can keep track of the regions it is requested.

In contrast, with this topology

reader -> resample -> monitor -> writerOrStreamingFilter

`monitor` does not know whether or not `resample` is streaming.

Have patch, will push :-)

**Comment by thewtex** [comment]
rebasing on `master`...

**Comment by romangrothausmann** [comment]
Cool. Many thanks @thewtex for taking a look and sorting things out. It wasn't easy for me to fully see your changes, i.e. those between my old https://github.com/InsightSoftwareConsortium/ITK/commit/29f4e00f6883b07119df55bc9b7565f03d51df9c and the new one https://github.com/InsightSoftwareConsortium/ITK/pull/392/commits/f9732088f9e81c7dc35f883e36c2df4a29f8a6c4 due to rebasing on newest master and squashing. 
So I rebased https://github.com/InsightSoftwareConsortium/ITK/commit/29f4e00f6883b07119df55bc9b7565f03d51df9c onto https://github.com/InsightSoftwareConsortium/ITK/commit/a1bd6835b28b85ba7e6e837ac0e177ae73016fa1 (parent of https://github.com/InsightSoftwareConsortium/ITK/pull/392/commits/f9732088f9e81c7dc35f883e36c2df4a29f8a6c4) and renamed to `itkResampleImageTest2Streaming.cxx` accordingly which results in https://github.com/romangrothausmann/ITK/commit/4566b7419b4c2bfa860a18957fd30db49eb26d55. 
Sadly, GH comparison does not work well in that case (https://github.com/InsightSoftwareConsortium/ITK/compare/f9732088f9e81c7dc35f883e36c2df4a29f8a6c4...romangrothausmann:4566b7419b4c2bfa860a18957fd30db49eb26d55) showing `itkResampleImageTest2Streaming.cxx`  and `itkResampleImageTest7.cxx` as new files instead of just the differences. So I create a new commit on top of https://github.com/romangrothausmann/ITK/commit/4566b7419b4c2bfa860a18957fd30db49eb26d55 with the files as in https://github.com/InsightSoftwareConsortium/ITK/pull/392/commits/f9732088f9e81c7dc35f883e36c2df4a29f8a6c4 resulting in https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe, so looking at that in GH should show the actual changes from @thewtex between https://github.com/InsightSoftwareConsortium/ITK/commit/29f4e00f6883b07119df55bc9b7565f03d51df9c https://github.com/InsightSoftwareConsortium/ITK/pull/392/commits/f9732088f9e81c7dc35f883e36c2df4a29f8a6c4 (please correct me if I got that wrong).

So there I can see that (apart from white space changes https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-50f53fc01f72ba938548d64a1218589bL79):
1. the unnecessary readers and their args got removed (https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-50f53fc01f72ba938548d64a1218589bL93).
2. typdefs replaced by using (sorry, I didn't catch those in the transition)
3. the order was changed to reader->monitor->resample (https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-50f53fc01f72ba938548d64a1218589bR121 and https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-50f53fc01f72ba938548d64a1218589bR175) which is great and (I understand now), same for `itkResampleImageTest7.cxx` (https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-dfa5983cf67253ce0e03aca56a207db2R91).
4. that `ClearPipelineSavedInformation` was introduced (e.g. https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-dfa5983cf67253ce0e03aca56a207db2R137)
5. VerifyAllInputCanStream was changed from 1 to 8: https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-50f53fc01f72ba938548d64a1218589bR181
6. `Modified()` is called on `image` instead of `resample` https://github.com/romangrothausmann/ITK/commit/e41fff83f86407feae9552354e4567eb7ca96efe#diff-dfa5983cf67253ce0e03aca56a207db2R137

I think all are fine except 5, where I intended that the monitor filter checks that streaming took not place, which I thought `VerifyAllInputCanStream(1)` would do. That and some minor corrections are proposed in #468.

Just for my better understanding: What is point 4. for, and why the change of 6.?

**Comment by romangrothausmann** [comment]
@thewtex Just noticed that for `itkResampleImageTest2Streaming` the comparison to the output of `itkResampleImageTest2` was removed: 
https://github.com/InsightSoftwareConsortium/ITK/compare/29f4e00f6883b07119df55bc9b7565f03d51df9c..69bb0bf11e345aaeeb4f72066d53b8d822061411#diff-e496b4f78280279f13a386cdbc983d41L255
Is that intended? My idea was that the `itkResampleImageTest2Streaming` __return value and the comparison__ with the output from `itkResampleImageTest2` should determine failure or success of the test.

**Comment by thewtex** [comment]
@romangrothausmann the comparison could be added back. If the test depends on the output of `itkResampleImageTest2`, we will need to add [`DEPENDS`](https://discourse.itk.org/t/itk-5-0-release-candidate-1-meshes/1576) property to ensure `itkResampleImageTest2` is executed first hen running in parallel.

**Pull request merged**